Fixed Rewards Onboarding UI #159

Brandon-T opened the pull request:
Fixes https://github.com/brave/brave-rewards-ios/issues/159

Completes the on-boarding flow for new and existing users.

## Submitter Checklist:

- [ ] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:
<!-- Any useful notes explaining how best to test and verify. -->


## Screenshots:
<!-- If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here. -->


## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/internal/issues/new?assignees=&labels=security%2C+security-reviews&template=security-review.md&title=iOS:%20Security+Review+for) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue is assigned to a milestone (should happen at merge time).

Brandon-T committed:
Added Transition Animations to the next screen..

Brandon-T committed:
Fixed bad animation glitches.

Brandon-T committed:
Fixed animations more. Added last screen in the flow.

Brandon-T committed:
Added existing user flow.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingWebViewController.swift:
```diff
@@ -0,0 +1,279 @@
+    
+    @objc
+    private func onAction() {
+        print("Access Button Pressed..")
```
What's this supposed to do?

Brandon-T replied:
No idea.. there is a button on the top right of the web-view that I have no information about.

This terms shows in its own web-view.

kylehickinson replied:
Oh, so this was a designed screen? I just thought you made it up 😆 I'd ping the designer who did it (Karen or James)

Brandon-T replied:
> No way to exit the T&C screen. A glitch after tapping start browsing and the the countdown looks a little blurred.
> ...
> ...

You can exit the terms and conditions by pressing the X button at the top left.. or by swiping down on the screen.

![image](https://user-images.githubusercontent.com/1530031/65168471-97a5c200-da12-11e9-9f8e-286fd47ccaf1.png)

danishjafri88 replied:
@Brandon-T as you can see in the screenshot I don't have the nav bar, nor is it modal. it shows full screen

danishjafri88 replied:
![ezgif com-resize-2](https://user-images.githubusercontent.com/5398557/65169214-ba13fb80-da63-11e9-8a5f-032963fe3838.gif)
iPhone XS iOS 12  (simulator & device)

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingViewController.swift:
```diff
@@ -18,6 +18,10 @@ class OnboardingViewController: UIViewController {
     }
     
+    var themeColour: UIColor {
+        return theme.isDark ? UIColor(rgb: 0x343A40) : UIColor(rgb: 0xFFFFFF)
```
Should these colors be defined somewhere?

Brandon-T replied:
They were already defined on development and I moved them to here instead.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingViewController.swift:
```diff
@@ -18,6 +18,10 @@ class OnboardingViewController: UIViewController {
         return theme.isDark
     }
     
+    var themeColour: UIColor {
```
This doesn't seem to update if dark mode is changed during the onboarding

Brandon-T replied:
None of the Brave UI seems to update when I switch from dark to light mode and vice-versa. I've been killing the browser to refresh it. This is true for development right now.

I only "moved" the code that already exists on development.

kylehickinson replied:
That's definitely not the case. I just launched the live version, and changed my phone to dark mode and it switches everything to dark right away

Brandon-T replied:
I tested it a million times. It does NOT switch for me on simulator. In any case, I copied the same code used to theme the rest of the application into the onboarding. Hopefully it works. On simulator it does NOT show for me. On iPhone-X it does.

I was wondering why this was left out in the original on-boarding code in the first place..

danishjafri88 commented:
No way to exit the T&C screen. A glitch after tapping start browsing and the the countdown looks a little blurred.
<img src="https://user-images.githubusercontent.com/5398557/65157247-153af380-da4e-11e9-9364-c4ad191f5a4e.jpeg" width="40%">                                                       <img src="https://user-images.githubusercontent.com/5398557/65158584-45839180-da50-11e9-9042-f6f35cad63fc.PNG" width="40%">
<img src="https://user-images.githubusercontent.com/5398557/65158857-c93d7e00-da50-11e9-895c-9d60ec403853.PNG" width="40%">

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownViewController.swift:
```diff
@@ -0,0 +1,71 @@
+        }
+    }
+    
+    private func generateNotification() {
```
This and all the UserNotifications related stuff should be removed and use the real my first ad stuff, which is already in `development`

Brandon-T replied:
Fixed.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownView.swift:
```diff
@@ -0,0 +1,370 @@
+        $0.alignmentMode = .center
+    }
+    
+    private let gradientLayer = { () -> CAGradientLayer in
```
Nit/Consistency: Switch all these to use `.then` i.e. 

```
private let gradientLayer = CAGradientLayer().then {
  $0.type = .conic
  // etc
}
```

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownView.swift:
```diff
@@ -0,0 +1,370 @@
+            }
+            
+            set {
+                countdownView.countdownLayer.string = newValue as NSString?
```
Is the cast really necessary?

Brandon-T replied:
Fixed.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownView.swift:
```diff
@@ -0,0 +1,370 @@
+                $0.edges.equalToSuperview().inset(UX.descriptionContentInset)
+            }
+            
+            [descriptionView].forEach(mainStackView.addArrangedSubview(_:))
```
No need to loop

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownView.swift:
```diff
@@ -0,0 +1,370 @@
+    func animate(from startOffset: CGFloat, to endOffset: CGFloat, duration: CFTimeInterval, completion: (() -> Void)? = nil) {
+        CATransaction.begin()
+        
+        countdownLayer.string = String(Int(duration)) as NSString
```
Is the cast necessary here?

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownView.swift:
```diff
@@ -0,0 +1,370 @@
+    }
+}
+
+private class CenteredTextLayer: CATextLayer {
```
Why do we need to use a CATextLayer instead of just a regular UILabel?

Brandon-T replied:
Can't animate UILabel's text property so I can't animate the countdown. If I use `CATextLayer`, I can animate the string property

kylehickinson replied:
What are we animating? I don't see any animation of the actual text in simulator, it just goes from 3 to 2 instantly? And its blurry like @danishjafri88 mentioned probably because it's a CATextLayer.  Unless I'm missing something, lets just replace it with a UILabel

Brandon-T replied:
Blurriness is fixed by setting the scale on the layer.

To keep things in sync with the rest of animations, it must be a CATextLayer. A UILabel's text cannot be animated so it won't be able to countdown with the rest of the animation, unless I add a timer and then manually change the text every tick.

Plus when timers go in the background they get expired so it's even worse than the current solution which works fine.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownViewController.swift:
```diff
@@ -0,0 +1,71 @@
+        completionHandler([.alert])
+        
+        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
+            self.contentView.setState(.adConfirmation)
```
When the user clicks the "My First Ad" it should open the my-first-ad URL as it does right now in the BVC. Are we going to be able to show this URL here? Or should we show it to them when they finally finish onboarding?

Brandon-T replied:
The UX shows that the ad shows on top of the onboarding screens and when you tap, the onboarding is complete and the BVC shows. If you don't tap, you get the `adConfirmation` screen to tell you why you might have missed the ad.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsFinishedView.swift:
```diff
@@ -0,0 +1,110 @@
+                $0.edges.equalToSuperview().inset(UX.descriptionContentInset)
+            }
+            
+            [descriptionView].forEach(mainStackView.addArrangedSubview(_:))
```
No need for loop

kylehickinson replied:
Not resolved for this one (accidental resolve or just missed it when pushing fixes?)

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsFinishedView.swift:
```diff
@@ -0,0 +1,110 @@
+            
+            [descriptionView].forEach(mainStackView.addArrangedSubview(_:))
+
+            [UIView.spacer(.horizontal, amount: 0),
```
I don't understand why we are using spacer views? Just use UIStackView's `setCustomSpacing`

Brandon-T replied:
This doesn't work. You can't set custom spacing "before". It also depends on the layout and since this layout requires "centering", a blank view with lower hugging before allows it to achieve equal centering and proper spacing. Without it, the UI is completely broken.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift:
```diff
@@ -127,3 +175,140 @@ extension OnboardingNavigationController {
+    }
+    
+    func transitionDuration(using transitionContext: UIViewControllerContextTransitioning?) -> TimeInterval {
+        
```
whitespace

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsAgreementView.swift:
```diff
@@ -0,0 +1,223 @@
+                $0.edges.equalToSuperview().inset(UX.descriptionContentInset)
+            }
+            
+            [descriptionView].forEach(mainStackView.addArrangedSubview(_:))
```
No need to loop

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsAgreementView.swift:
```diff
@@ -0,0 +1,223 @@
+            
+            [descriptionView].forEach(mainStackView.addArrangedSubview(_:))
+
+            [skipButton, agreeButton, UIView.spacer(.horizontal, amount: 0)]
```
Same comment about spacer views as above → use `UIStackView.setCustomSpacing(_:after:)`

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsAgreementViewController.swift:
```diff
@@ -0,0 +1,28 @@
+    }
+    
+    override func loadView() {
+        view = View(theme: theme, themeColour: themeColour)
```
I would sync up with @iccub on the way we're supposed to do theming for onboarding

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsView.swift:
```diff
@@ -0,0 +1,132 @@
+                $0.edges.equalToSuperview().inset(UX.descriptionContentInset)
+            }
+            
+            [descriptionView].forEach(mainStackView.addArrangedSubview(_:))
```
No need for loop

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsView.swift:
```diff
@@ -0,0 +1,132 @@
+            
+            [descriptionView].forEach(mainStackView.addArrangedSubview(_:))
+
+            [skipButton, joinButton, UIView.spacer(.horizontal, amount: 0)]
```
No spacers, same as above

Brandon-T replied:
All of these spacers are currently in dev. I just copied all of them to keep the code consistent and keep the UI the exact same as what's already there: https://github.com/brave/brave-ios/search?utf8=%E2%9C%93&q=UIView.spacer%28.vertical&type=

Even on the old on-boarding code it is there.

Removing them completely breaks the UI as the stackView is centered and the spacer has lower content hugging priority than other views so it takes up the remaining space allowing the desired views to be centered properly and spaced equally.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingShieldsView.swift:
```diff
@@ -67,38 +67,44 @@ extension OnboardingShieldsViewController {
                 $0.edges.equalToSuperview().inset(UX.descriptionContentInset)
             }
+            
+            [descriptionView].forEach(mainStackView.addArrangedSubview(_:))
```
No need to loop

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingShieldsView.swift:
```diff
@@ -67,38 +67,44 @@ extension OnboardingShieldsViewController {
+            
+            [descriptionView].forEach(mainStackView.addArrangedSubview(_:))
+            
+            [skipButton, continueButton, UIView.spacer(.horizontal, amount: 0)]
```
No spacers

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingWebViewController.swift:
```diff
@@ -0,0 +1,279 @@
+    }
+    
+    private func updateBackForwardUI() {
+        toolbar.backButton.isEnabled = webView.canGoBack
```
Remove all the delegate methods below that are implemented simply to update UI and just subscribe to relevant KVO as BrowserViewController does

Brandon-T replied:
Fixed.

danishjafri88 reviewed (commented):
@kylehickinson beat me to the review :). Just one question, should the onboarding not show if the `NO_REWARDS` flag is set?

kylehickinson commented:
@danishjafri88 You're right, it probably shouldn't since you can't create a BraveRewards to do it.

@Brandon-T  Made me realize that this doesn't actually enable rewards (create a wallet) when you accept the terms like it should... So add that to the list 😛

You will also need to add a loader to the button which creates the wallet (check the rewards panel)

danishjafri88 commented:
Another issue (Not very crucial imo): If the user backgrounds the app while the counter is running and he comes back after 3 seconds the timer will not update. This is because the CALayer animations pause on background. Maybe pause and resume, or put a timer to update ui as well?
![ezgif com-resize](https://user-images.githubusercontent.com/5398557/65163340-62bc5e00-da58-11e9-84a0-9ac99d9502a0.gif)

Brandon-T committed:
Addressed Feedback.

Brandon-T committed:
Addressed Feedback and generate my-first-ad with BVC.

Brandon-T committed:
Use KVO to update webview UI.

Brandon-T committed:
Fixed glitchy animations.

kylehickinson reviewed (changes requested):
As mentioned: Rewards has to actually be enabled and turned on before the user can go to the "waiting for first ad" screen

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift:
```diff
@@ -53,13 +65,25 @@ class OnboardingNavigationController: UINavigationController {
             case .shieldsInfo: return OnboardingShieldsViewController.self
+            case .rewardsInfo: return OnboardingRewardsViewController.self
+            case .rewardsAgreement: return OnboardingRewardsAgreementViewController.self
+            case .adsCountdown: return OnboardingAdsCountdownViewController.self
```
I think as per spec we shouldn't be showing the ads stuff unless the user can actually enable ads (`BraveAds.isSupportedRegion(Locale.current.identifier)` returning true)

Brandon-T replied:
Fixed.

kylehickinson commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -245,7 +245,7 @@ class BrowserViewController: UIViewController {
     
     // Display first ad when the user gets back to this controller if they havent seen one before
-    func displayMyFirstAdIfAvailable() {
+    func displayMyFirstAdIfAvailable(_ completion: (() -> Void)? = nil) {
```
This change doesn't do anything? `completion` isn't called.

Either way, lets remove this change and simply copy/paste similar code into onboarding (barring figuring out how to show the my-first-ad URL).

Brandon-T replied:
Fixed.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownViewController.swift:
```diff
@@ -0,0 +1,46 @@
+        contentView.animate(from: 0.0, to: 1.0, duration: 3.0) { [weak self] in
+            guard let self = self else { return }
+            
+            (UIApplication.shared.delegate as? AppDelegate)?.browserViewController.displayMyFirstAdIfAvailable({
```
This won't actually show up on onboarding because its presented on the BVC (as talked about above)

Brandon-T replied:
Fixed. Max copy paste duplicate code..

Brandon-T committed:
Fixed for-loop feedback.

Brandon-T committed:
Added Darkmode dynamic toggling to switch between dark and light at runtime.

Brandon-T committed:
Added terrible amounts of copy-paste.. Added ifdef to determine if to enable rewards.

Brandon-T committed:
Added check for BraveAds region.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownViewController.swift:
```diff
@@ -0,0 +1,128 @@
+        rewardsObserver = nil
+        #else
+        RewardsHelper.configureRewardsLogs()
+        rewards = BraveRewards(configuration: .default)
```
Hmmm, none of this is needed at all in onboarding. All we need is 1. to check that BraveAds is supported in this region, which doesn't require us make a new one, and 2. the `AdsViewController.displayFirstAd` call

In fact, the only place we need an actual rewards instance is when clicking "accept" or w/e the button is where we need to create the wallet and enable rewards (which I think this PR is still missing). This should be passed _into_ the onboarding controller since the BVC already makes that object (which you copied here)

Is there a reason you added all this and the rewards observers and everything?

Brandon-T replied:
Lol.. why does it seem like this is more complicated than it needs to be? I just need to show a notification and know when it has shown and if the user tapped it.

I already have the check for if ads are supported in the NavigationController which determines what screens will be shown. So that's done.

Anyway, it is all there because I need to know if the notification has shown, and if the user tapped it, through the use of `AdsNotificationHandler(ads: rewards.ads, presentingController: self)` which has a rewards.ads parameter.. The code is so tightly coupled I can't have one without the other. To get notified, I have to create a BraveRewards and then a NotificationHandler and then display the notification. But now I need a wallet before all of this which can only be done through BVC? OR does the callback for `displayFirstAd` notify me of the user's actions? Did I understand this correct?

If this is correct, I will add it now but it isn't clear. 

BVC is going to pass the onboarding navigation controller a `BraveRewards` instance 
 and it will then pass that to subsequent child onboarding controllers to generate a `notification` and do all the same logic that BVC already does?

kylehickinson replied:
The `AdsNotificationHandler` is completely separate from the "My First Ad".  You know it showed up based on whether or not it was dismissed/timed out/was clicked/etc from the AdsViewController displayFirstAd method. Also the `notificationAdded` observer method isn't for ads, it's for the wallet notifications (insufficient funds, grant added, etc.)

The loading part, means the whole point of this onboarding is to get users to opt-in to Brave Rewards. To opt-in you have to create a wallet (like the panel does). Simply showing them a first ad is nothing but a special UI.

The spec for this is in Drive: `User Onboarding iOS - Phase 2` under **Cards**, point 3 & 4

kylehickinson replied:
The code related to creating a wallet is in brave-rewards-ios: 
https://github.com/brave/brave-rewards-ios/blob/0f598b794003adc7231e6059b6f075ca798314cc/BraveRewardsUI/Extensions/BraveLedgerExtensions.swift#L89

This may have to become public though, otherwise its the same functionality

Brandon-T replied:
This can't be right. I need to know if the ad shown as well.. not just if the user tapped it.
So that I can show the "Did you see an ad screen" or take them directly to the browser.. but it's fine I'll just remove that and show it always then.

Creating the wallet part is fine.. It will be in the next commit.

kylehickinson replied:
If you need additional stuff from the My First Ad display the code for it is here and could be modified to support more than just tapping:

https://github.com/brave/brave-rewards-ios/blob/5fe6a61bd8735308bd6f079b21938a038fe22301/BraveRewardsUI/Ads/AdsViewController.swift#L308

There's nothing really to notify about "this was shown". We just assume it was shown because we're adding it to the view hierarchy ourselves. Maybe we could add a callback for the "animate out" part and if the user didn't tap it before it disappeared then we show the did you see it part, but just adding more complexity

kylehickinson replied:
Just basically assume "My first ad" has nothing to do with real ads or the setup/logic around ads, because it doesn't function like it at all other than my first ad showing the same UI

Brandon-T replied:
Done. Code is added to create a wallet. It is commented out for now until the interface is public.

Added loader when the user agrees. Removed all the rewards observer code and kept only ad creation code.

kylehickinson replied:
Okay so make the changes in brave-rewards-ios and update the Carthage dep here (Cartfile.resolved update commited) when it's done and uncomment and will continue review then

kylehickinson replied:
If you need additional updates for the first ad stuff probably best to do them at the same time first in brave-rewards-ios and then you can fix both when the dep is updated in this PR

Brandon-T replied:
Done.

Brandon-T committed:
Added ability to create a wallet.. Commented out for now since the interface is not public.
Removed all the rewards notification code..

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsAgreementViewController.swift:
```diff
@@ -0,0 +1,56 @@
+        }
+        
+        loadingView.startAnimating()
+        rewards?.ledger.createWallet { [weak self] error in
```
Should be `createWalletAndFetchDetails`

kylehickinson replied:
Oh, and we should guard this from being able to be tapped multiple times (Check whether or not loading view is already animating, or a bool flag for instance)

kylehickinson reviewed (changes requested):
Needs updated brave-rewards-ios dev (Cartfile.resolved file)

Brandon-T committed:
Fixed minors.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsAgreementViewController.swift:
```diff
@@ -0,0 +1,61 @@
+            self.loadingView.stopAnimating()
+            self.loadingView.removeFromSuperview()
+            self.contentView.agreeButton.setTitleColor(titleColour, for: .normal)
+            self.continueTapped()
```
Shouldn't we check for an error here before automatically moving them on?

Brandon-T committed:
Added error handling for when wallet creation fails.

kylehickinson reviewed (changes requested):
Found some bugs while runtime testing:
1. After creating a wallet within onboarding, the Brave Rewards icon doesn't work. Tapping it yields no panel popup. Restarting app fixes it
2. As per Slack, selecting "I didn't see an ad" points to a screen which doesn't make sense in our case
3. The circle for "your ad will appear in…" is blurry and could use the `scale` treatment that the text layer got
4. The ••• icon in the web view still doesnt do anything. (Edit: as per Slack, to be removed)
5. The first ad appears incorrectly on iPad: ![IMG_0106](https://user-images.githubusercontent.com/529104/65281365-74594080-db00-11e9-9eb2-0f4c92d2cf98.jpg)

Brandon-T committed:
Removed extra screens that no longer apply.

Brandon-T committed:
Removed action button from the webview.

Brandon-T committed:
Update carthage.

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownView.swift:
```diff
@@ -0,0 +1,390 @@
+                buttonsStackView.isHidden = true
+                countdownView.isHidden = false
+                titleLabel.text = Strings.OBAdsTitle
+                break
```
Remove break

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownView.swift:
```diff
@@ -0,0 +1,390 @@
+                buttonsStackView.isHidden = false
+                countdownView.isHidden = true
+                titleLabel.text = Strings.OBCompleteTitle
+                break
```
remove break

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownView.swift:
```diff
@@ -0,0 +1,390 @@
+            [titleLabel, countdownView, buttonsStackView].forEach(descriptionStackView.addArrangedSubview(_:))
+            
+            countdownView.snp.makeConstraints {
+                $0.width.equalTo(156.0)
```
I believe you can use snapkit's `size` property to set both width and height if they are the same.
This constant could also go in `UX` struct

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownView.swift:
```diff
@@ -0,0 +1,390 @@
+        
+        func applyTheme(_ theme: Theme) {
+            descriptionView.backgroundColor = OnboardingViewController.colourForTheme(theme)
+            titleLabel.textColor = theme.isDark ? .white : .black
```
Change `textColor` to `appearanceTextColor` for proper theming on iOS 12.

Also replace `.white` and `.black` with `theme.colors.tints.home`, and please test if it works correctly(UI looks good)

kylehickinson reviewed (changes requested):
We should disable the "skip" button while it is creating a wallet during the Brave Rewards step

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownView.swift:
```diff
@@ -0,0 +1,390 @@
+    override func layoutSubviews() {
+        super.layoutSubviews()
+
+        //Layout shapeLayer
```
In general, we add a space after a comment block

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownViewController.swift:
```diff
@@ -0,0 +1,83 @@
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        
+        if let time = timeSinceAnimationStarted, Date().timeIntervalSince(time) >= 3.0 {
```
I think we should extract this `3` number to a integer constant and then cast it to string or double, as we use it in few places here

kylehickinson commented:
Also the animation circle thing is still blurry despite the content scale lines being added

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsAgreementViewController.swift:
```diff
@@ -0,0 +1,69 @@
+        super.viewDidLoad()
+        
+        contentView.agreeButton.addTarget(self, action: #selector(onAgreed), for: .touchDown)
+        contentView.skipButton.addTarget(self, action: #selector(skipTapped), for: .touchDown)
```
These should be `touchUpInside` instead of `touchDown`

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownViewController.swift:
```diff
@@ -0,0 +1,83 @@
+        
+        //Countdown timer
+        contentView.countdownText = "3"
+        contentView.finishedButton.addTarget(self, action: #selector(continueTapped), for: .touchDown)
```
These should be `touchUpInside` instead of `touchDown`

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift:
```diff
@@ -19,6 +22,12 @@ protocol OnboardingControllerDelegate: class {
 
+enum OnboardingViewAnimationID: Int {
+    case background = 1
+    case details = 2
```
you don't need to declare `2,3...` if you have assigned `1` to first case, it's inferred
You can leave it if you find it more readable though

kylehickinson commented:
> After creating a wallet within onboarding, the Brave Rewards icon doesn't work. Tapping it yields no panel popup. Restarting app fixes it

This bug is also still happening.

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsAgreementView.swift:
```diff
@@ -0,0 +1,231 @@
+            stackView.spacing = 8
+            let descriptionStackView =  UIStackView(arrangedSubviews: [descriptionLabel, descriptionCheckbox]).then {
+                $0.alignment = .center
+                $0.spacing = 65.0
```
Will this look good on all devices?
I guess the goal here is to have description aligned to left and checkbox aligned to right?

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsAgreementView.swift:
```diff
@@ -0,0 +1,231 @@
+            descriptionLabel.attributedText = {
+                let text = NSMutableAttributedString(string: Strings.OBRewardsAgreementDetail, attributes: [
+                    .font: UIFont.systemFont(ofSize: 16, weight: UIFont.Weight.regular),
+                    .foregroundColor: titleLabel.textColor!
```
we could unwrap it at the beginning of the method, it's used in few places here

or make something with a fallback like `let color = titleLabel.textColor : .black`

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsAgreementView.swift:
```diff
@@ -0,0 +1,231 @@
+        
+        func applyTheme(_ theme: Theme) {
+            descriptionView.backgroundColor = OnboardingViewController.colourForTheme(theme)
+            titleLabel.textColor = theme.isDark ? .white : .black
```
same as other comment, use `appearanceTextColor` and `theme.color.tints.home`

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsAgreementView.swift:
```diff
@@ -0,0 +1,231 @@
+            titleLabel.textColor = theme.isDark ? .white : .black
+            updateDescriptionLabel()
+            if theme.isDark {
+                descriptionCheckbox.setImage(#imageLiteral(resourceName: "checkbox_off_dark"), for: .normal)
```
can be simplified to `let image = theme.isDark ? img1 : img2`
and then set the image

Brandon-T commented:
> > After creating a wallet within onboarding, the Brave Rewards icon doesn't work. Tapping it yields no panel popup. Restarting app fixes it
> 
> This bug is also still happening.


https://i.imgur.com/O0ND3um.mp4

Works for me 100% of the time..
That button is also not part of the onboarding flow, so I can't see how it affects it in any way.

For the blur part, I added rasterization scaling but I'm not seeing any blur on my end so if it doesn't fix it, I'm out of ideas. However, according to S/O this is the solution.

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsAgreementView.swift:
```diff
@@ -0,0 +1,231 @@
+        }
+        
+        @objc
+        func textView(_ textView: UITextView, shouldInteractWith URL: URL, in characterRange: NSRange, interaction: UITextItemInteraction) -> Bool {
```
Could we move it and `UITextViewDelegate` to an extension below?

Brandon-T committed:
Addressed Feedback.

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsView.swift:
```diff
@@ -0,0 +1,137 @@
+        
+        func applyTheme(_ theme: Theme) {
+            descriptionView.backgroundColor = OnboardingViewController.colourForTheme(theme)
+            titleLabel.textColor = theme.isDark ? .white : .black
```
see my other comments about this

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsView.swift:
```diff
@@ -0,0 +1,137 @@
+        }
+        
+        func applyTheme(_ theme: Theme) {
+            descriptionView.backgroundColor = OnboardingViewController.colourForTheme(theme)
```
rename `colourForTheme` to `colorForTheme`

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingSearchEnginesView.swift:
```diff
@@ -120,6 +130,14 @@ extension OnboardingSearchEnginesViewController {
+        func applyTheme(_ theme: Theme) {
+            containerView.backgroundColor = OnboardingViewController.colourForTheme(theme)
+            searchEnginesTable.backgroundView?.backgroundColor = OnboardingViewController.colourForTheme(theme)
+            titlePrimary.textColor = theme.isDark ? .white : .black
```
see other comment about theming this

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingShieldsView.swift:
```diff
@@ -67,38 +73,52 @@ extension OnboardingShieldsViewController {
+            descriptionView.backgroundColor = OnboardingViewController.colourForTheme(theme)
+            textStackView.arrangedSubviews.forEach({
+                if let label = $0 as? UILabel {
+                    label.textColor = theme.isDark ? .white : .black
```
see other comment about theming this

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingViewController.swift:
```diff
@@ -5,21 +5,37 @@
     }
     
-    init(profile: Profile, theme: Theme) {
+    static func colourForTheme(_ theme: Theme) -> UIColor {
```
rename to `color` without `u`

iccub reviewed (changes requested):
Just leaving some smaller feedback and code style concerns.
Kyle is doing good job reviewing this PR

kylehickinson commented on Cartfile.resolved:
```diff
@@ -3,8 +3,8 @@ github "DaveWoodCom/XCGLogger" "7.0.0"
 github "SwiftyJSON/SwiftyJSON" "5.0.0"
-github "airbnb/lottie-ios" "531eb22d47a9439a2344fdaf96832189f2c6e925"
-github "brave/brave-rewards-ios" "092cd4e52d9d821c7186de941adc6be8126baf23"
+github "airbnb/lottie-ios" "860124f1e8fe2163a87c225a931132fc57324b26"
```
Is this change to Lottie library intentional?

Brandon-T replied:
When I updated carthage, it also updated. If we don't want it, I can remove it but it works fine.

kylehickinson replied:
Probably best to double check what changed in the update on `airbnb/lottie-ios` side.

Also you can update a single lib by passing the name into carthage update like: `carthage update brave-rewards-ios <other options>`

Brandon-T committed:
Addressed Feedback.

Brandon-T committed:
Resolved Lottie to the correct commit.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownViewController.swift:
```diff
@@ -0,0 +1,87 @@
+        if Preferences.Rewards.myFirstAdShown.value { return }
+
+        if BraveAds.isSupportedRegion(Locale.current.identifier) {
+            DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
```
This 3 second delay is not needed, since you already wait until the animation completes which is 3 seconds long

kylehickinson replied:
@Brandon-T You marked this resolved but I don't see the change… unresolving for now

Brandon-T replied:
Hmm for some reason it didn't go through when I committed via Xcode. Should be there this time.

kylehickinson commented:
Missing flows from `Existing User Onboarding iOS` spec in Drive:

- User has already seen all onboarding, rewards are off and ads become available in users region → User should be shown the onboarding for just rewards opt-in and my first ad
- User has already seen regular onboarding, and ads are available in users region → User should be shown the onboarding for just rewards opt-in and my first ad

kylehickinson commented:
Well I tested it after changing my region back to US and it didn't open onboarding

Brandon-T committed:
Added missing flow where user has seen onboarding before and ads became available in their region.

Brandon-T commented:
> Well I tested it after changing my region back to US and it didn't open onboarding

I just noticed.. the code had a "TBD" in it from the previous developer. I added the extra flows.

Brandon-T committed:
Fixed logic in new flows.

kylehickinson commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -679,10 +679,11 @@ class BrowserViewController: UIViewController {
             onboarding.onboardingDelegate = self
             present(onboarding, animated: true)
         } else if BraveAds.isSupportedRegion(Locale.current.identifier) {
+            let isAdsEnabled = rewards?.ads.isEnabled ?? false
```
Is this to check if rewards is off? If so, I'd do `rewards?.ledger.isEnabled == false`

Brandon-T committed:
Addressed Feedback.

kylehickinson commented:
Just have to get verification on what to do when the user does have Rewards enabled already and then ads becomes available in their region. At the moment enabling Rewards enables ads (just as it did in Desktop where I copied the code from), but the spec says the onboarding should come up again, which means we'd need to change it to only enable ads by default if their region is supported. And then this second onboarding would just turn on ads, not enable rewards, which doesn't require ToS or anything (cc @anthonypkeane) 

Probably will have further discussion

Brandon-T committed:
Update carthage.

Brandon-T committed:
Onboarding Added.
Added animation for Rewards & Ads.
Added logic for transitioning between each onboarding screen.

Brandon-T committed:
Changed Brave Rewards screen to reflect the new design text.
Changed Terms of Agreement screen to allow "cancel" which takes you backwards to the Brave Rewards screen.
Changed countdown to 5 seconds instead of 3 which was too quick.
Added flow for when 60 days passes and the user skipped the onboarding.

Brandon-T committed:
Added flow for when the user skipped the onboarding (60-days).

Brandon-T committed:
Changed everything to a state-machine. Way easier to handle the change in flows and understand what's happening..

kylehickinson commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -667,17 +669,81 @@ class BrowserViewController: UIViewController {
+                return
+            }
+            
+            // 60 days has passed since the user last saw the onboarding.. it's time to show the onboarding again..
```
Is the 60 days for basic onboarding as well?

Brandon-T replied:
It's actually not clarified if it's for basic or not but it does say:

"If the user SKIPS" the on-boarding, we show them again in 60 days. I'm not sure if this applies to search engine but I can't see why it won't since it's also onboarding. I'll ask.

kylehickinson replied:
This matches Android, and we will be sticking with that for now

kylehickinson commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -667,17 +669,81 @@ class BrowserViewController: UIViewController {
+        
+        // 1. Rewards are on/off (existing user)
+        // 2. User hasn't seen the rewards part of the onboarding yet.
+        if Preferences.General.basicOnboardingCompleted.value == OnboardingState.completed.rawValue && Preferences.General.basicOnboardingProgress.value == OnboardingProgress.searchEngine.rawValue {
```
Lol these if statements are getting long, can we move each condition onto a new line 😂

Brandon-T replied:
Fixed.

kylehickinson commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -667,17 +669,81 @@ class BrowserViewController: UIViewController {
+        // 1. Rewards are on/off (existing user)
+        // 2. Ads are now available
+        // 3. User hasn't seen the ads part of onboarding yet
+        if BraveAds.isSupportedRegion(Locale.current.identifier) && Preferences.General.basicOnboardingCompleted.value == OnboardingState.completed.rawValue && Preferences.General.basicOnboardingProgress.value != OnboardingProgress.ads.rawValue {
```
Same as above

kylehickinson commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -667,17 +669,81 @@ class BrowserViewController: UIViewController {
+                onboarding.onboardingDelegate = self
+                present(onboarding, animated: true)
+                
+                Preferences.General.basicOnboardingNextOnboardingPrompt.value = Date(timeIntervalSinceNow: BrowserViewController.onboardingDaysInterval * 24 * 60 * 60)
```
I think there's some shared code that lets us do like `60.toDays` or something, can you double check that exists

Brandon-T replied:
Fixed.

kylehickinson commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -3104,9 +3170,19 @@ extension BrowserViewController: OnboardingControllerDelegate {
+    
+    func onboardingSkipped(_ onboardingController: OnboardingNavigationController) {
+        Preferences.General.basicOnboardingCompleted.value = OnboardingState.skipped.rawValue
+        Preferences.General.basicOnboardingNextOnboardingPrompt.value = Date(timeIntervalSinceNow: BrowserViewController.onboardingDaysInterval * 24 * 60 * 60)
```
Same as above

kylehickinson commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -3104,9 +3170,19 @@ extension BrowserViewController: OnboardingControllerDelegate {
+    }
+    
+    // 60 days until the next time the user sees the onboarding..
+    static let onboardingDaysInterval: Double = 60.0
```
Or maybe even just turn this into `60.days` or w/e it is and don't do any conversion above

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownViewController.swift:
```diff
@@ -0,0 +1,97 @@
+class OnboardingAdsCountdownViewController: OnboardingViewController, UNUserNotificationCenterDelegate {
+    
+    private struct UX {
+        static let animationTime = 5.0
```
As mentioned in Slack, this should be 3 here, and we'll update ad automatic dismissal duration elsewhere

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingAdsCountdownViewController.swift:
```diff
@@ -0,0 +1,97 @@
+    }
+    
+    override func continueTapped() {
+        Preferences.Rewards.myFirstAdShown.value = true
```
We should probably set this after displaying the view, incase the user for some reason force quits the app before clicking continue or something stupid like that and then it'd show again if they ever disabled/reenabled Brave Rewards

Brandon-T replied:
This part is weird. If I don't do this and they force quit before completing the onboarding, onboarding is technically no longer completed.. So it will popup the next time they launch in order for them to complete it.. However, they will NEVER see the popup because of this boolean.. Hence why I put it only when they press continue.

kylehickinson replied:
Ah I see… so if they force-quit in the middle of onboarding they'd see onboarding again?

Brandon-T replied:
Yeah. This is what was originally written before I took over the PR. Since they force quit, the onboarding state was never updated so they need to actually finish it or skip it if they don't want to see it.

```
/// If user kills the app before completing onboarding, it should be treated as unfinished.
static let basicOnboardingCompleted = Option<Int>(key: ...)
```

kylehickinson replied:
Gotcha, ok this is fine then

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift:
```diff
@@ -30,59 +42,90 @@ class OnboardingNavigationController: UINavigationController {
         preferredContentSize = UX.preferredModalSize
+        
+        let backgroundView = UIView().then {
+            $0.backgroundColor = #colorLiteral(red: 0.1176470588, green: 0.1254901961, blue: 0.1607843137, alpha: 1)
```
This color usage needed? Not sure if we replace it based on theme?

Brandon-T replied:
It's needed. It was there before my time on Onboarding. I just "moved" it out of each controller into just one parent controller.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingViewController.swift:
```diff
@@ -87,4 +108,17 @@ class OnboardingViewController: UIViewController {
         }
     }
+    
+    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
+        if #available(iOS 13.0, *) {
+            if UITraitCollection.current.userInterfaceStyle != previousTraitCollection?.userInterfaceStyle {
+                theme = OnboardingViewController.getUpdatedTheme()
+                applyTheme(theme)
+            }
+        }
+    }
```
@iccub / @jhreis This OK? I'm not super familiar with the theming work

Brandon-T replied:
This is how it's done everywhere else if you search for `applyTheme`

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingViewController.swift:
```diff
@@ -5,21 +5,37 @@
+            return Theme.from(id: isDark ? Theme.DefaultTheme.dark.rawValue : fallback)
+        }
+        return Theme.from(id: fallback)
+    }
```
@iccub @jhreis Same here

Brandon-T replied:
This code is copy pasted into onboarding from 

```
static func of(_ tab: Tab?) -> Theme {
```

in order to get the most recent theme info.. This is done because `Theme` is a struct (copy) and does NOT update automatically (class reference) if the user goes and switches the system to light/dark mode.

**EDIT**: I changed it to `Theme.of` which does the exact same thing so I don't duplicate code..

Brandon-T committed:
Addressed feedback.

Brandon-T committed:
Replaced theme logic with `Theme.of` instead. (Copied from TabTrayController and BrowserViewController).

Brandon-T committed:
Updated rewards.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift:
```diff
@@ -30,59 +42,90 @@ class OnboardingNavigationController: UINavigationController {
-            case .newUser: return [.searchEnginePicker, .shieldsInfo, /* .rewardsInfo, .adsInfo */]
-            case .existingUser: return [/* .rewardsInfo, .adsInfo */]
+            case .newUser: return [.searchEnginePicker, .shieldsInfo]
+            case .existingUser: return []
```
FYI: This line does not compile

`Type 'OnboardingNavigationController.OnboardingType' has no member 'existingUser'`

kylehickinson commented:
Can we increase the hit box on the checkbox when accepting ToS for Rewards? It's a bit hard to hit.

anthonypkeane commented:
> Can we increase the hit box on the checkbox when accepting ToS for Rewards? It's a bit hard to hit.

HIG is 44px, any hit space around the box from 32-44 should be fine.

kylehickinson reviewed (changes requested):
Hit this weird state, not sure if its because I've had it installed in the past or what, but it went to the "ads available" flow, and when clicked get started it went here but no ad shows up. Maybe we should just ignore check for "my first ad" when going through this countdown page? If the user somehow does end up on it I'd rather then see something. Also in this state there's no way to get out, since no ad appears, therefore you have no automatic dismissal to trigger the next step

![Simulator Screen Shot - iPhone 11 Pro - 2019-10-02 at 16 40 28](https://user-images.githubusercontent.com/529104/66080088-5d204700-e533-11e9-8f59-18eb14f1429d.png)

kylehickinson commented:
~Also this should say "Get started" instead of "Join" (because you aren't joining anything). I know the flow says join but I believe we said otherwise on the call~ Sticking with Join as below

anthonypkeane commented:
> Also this should say "Get started" instead of "Join" (because you aren't joining anything). I know the flow says join but I believe we said otherwise on the call
> 


Lets leave this as-is for now and test on the device. I know @karenkliu had good reason for that word choice.

iccub commented on Client.xcodeproj/project.pbxproj:
```diff
@@ -1692,13 +1701,22 @@
 		5E4845C122DE3DF800372022 /* WindowRenderHelperScript.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = WindowRenderHelperScript.swift; sourceTree = "<group>"; };
 		5E9288C922DF864C007BE7A6 /* TabSessionTests.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = TabSessionTests.swift; sourceTree = "<group>"; };
 		5EAB745B230456F0006055F5 /* BookmarkValidation.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = BookmarkValidation.swift; sourceTree = "<group>"; };
-		5EB57D0122FDC0CB00A07325 /* FullscreenHelper.js */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.javascript; path = FullscreenHelper.js; sourceTree = "<group>"; };
```
This file keeps getting removed in all PRs for some reason, should it be removed here?

Brandon-T replied:
Not sure why.. but if I ever rebase on development, it automatically does this. I have reverted it.

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift:
```diff
@@ -127,3 +197,156 @@ extension OnboardingNavigationController {
+        tDetails?.alpha = 0.0
+        tDetailsContent?.alpha = 0.0
+        
+        //guard let tDetailsFrame = tDetails?.superview?.convert(tDetails?.frame ?? .zero, to: container) else { return }
```
Remove comment if not needed

iccub commented on Client/Frontend/Browser/Onboarding/OnboardingNavigationController.swift:
```diff
@@ -127,3 +197,156 @@ extension OnboardingNavigationController {
+        //Setup
+        fromView.frame = container.bounds
+        toView.frame = container.bounds
+        //isPresenting ? container.addSubview(toView) : container.insertSubview(toView, belowSubview: fromView)
```
Remove comment if not needed

iccub reviewed (changes requested):
Small review + CI is still failing

Brandon-T committed:
Addressed feedback. Made terms and conditions checkbox hit area larger.

Brandon-T committed:
Fixed the state where the user can get stuck on the ads flow if they already saw it and it comes up again due to previous install.

Brandon-T committed:
Redo the entire checkbox insets to just use hit test of its parent.. UI got broken on SE device..

Brandon-T committed:
Updated to have real code instead of demo code. Tested to make sure the first ad really does show up and loading and different states work correctly.

kylehickinson commented on Client/Frontend/Browser/Onboarding/OnboardingRewardsAgreementView.swift:
```diff
@@ -0,0 +1,249 @@
+            agreeButton.isEnabled = button.isSelected
+        }
+        
+        override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
```
Why not just override `pointInside` instead of hit test:

```swift
override func point(inside point: CGPoint, with event: UIEvent?) -> Bool {
  if UIEdgeInsetsInsetRect(bounds, UX.checkboxInsets)).contains(point) {
    return true
  }
  return super.point(inside: point, with: event)
}
```

Edit: Just realized this is hitTest on the main view, not the checkbox, so point inside won't work

Brandon-T replied:
In case we are wondering. I HAVE to add it on the main view and not the button itself because the stackView is legit blocking the input to the button on the top and bottom (it clips it).. but NOT the sides due to the label next to it lol.

kylehickinson replied:
Yeah stack view's a bitch for not accepting touches outside its bounds :P

kylehickinson commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -255,9 +255,10 @@ class BrowserViewController: UIViewController {
             DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
                 if Preferences.Rewards.myFirstAdShown.value { return }
                 Preferences.Rewards.myFirstAdShown.value = true
-                AdsViewController.displayFirstAd(on: self) { [weak self] url in
-                    let request = URLRequest(url: url)
-                    self?.tabManager.addTabAndSelect(request, isPrivate: PrivateBrowsingManager.shared.isPrivateBrowsing)
```
This change is on purpose, coming from #1597, looks like was an accident on rebase?

Brandon-T replied:
It's on purpose. That other code is missing an argument for `displayFirstAd`

kylehickinson replied:
Yeah sorry I mean the part where it uses `addTabAndSelect` instead of `openNewTab`

Brandon-T committed:
For the sake of consistency.

karenkliu commented:
@Brandon-T @kylehickinson @anthonypkeane Caught up with Kyle about the "Join" button on the screen for Existing user BR on, Ads become available case. When I originally came up with the design, I thought that the user had to turn on ads when ads became available in their region. However, since ads are turned on by default when the user check BR ToS, we should change the button to "Show Me". That screen is purely to introduce users to ads; there is nothing to "turn on" or "join."
